core/trino-main/src/main/java/io/trino/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -996,16 +1002,28 @@ public PlanAndMappings visitJoin(JoinNode node, UnaliasContext context)
             Optional<Symbol> newRightHashSymbol = node.getRightHashSymbol().map(mapper::map);
 
             // rewrite dynamic filters
-            Set<Symbol> added = new HashSet<>();
+            Map<Symbol, DynamicFilterId> canonicalDynamicFilters = new HashMap<>();
             ImmutableMap.Builder<DynamicFilterId, Symbol> filtersBuilder = ImmutableMap.builder();
+            ImmutableMap.Builder<DynamicFilterId, DynamicFilterId> dynamicFilterIdMapBuilder = ImmutableMap.builder();
             for (Map.Entry<DynamicFilterId, Symbol> entry : node.getDynamicFilters().entrySet()) {
                 Symbol canonical = mapper.map(entry.getValue());
-                if (added.add(canonical)) {
+                DynamicFilterId canonicalDynamicFilterId = canonicalDynamicFilters.putIfAbsent(canonical, entry.getKey());
+                if (canonicalDynamicFilterId == null) {
                     filtersBuilder.put(entry.getKey(), canonical);
                 }
+                else {
+                    dynamicFilterIdMapBuilder.put(entry.getKey(), canonicalDynamicFilterId);
+                }
             }
             Map<DynamicFilterId, Symbol> newDynamicFilters = filtersBuilder.build();
 
+            ImmutableMap<DynamicFilterId, DynamicFilterId> dynamicFilterIdMap = dynamicFilterIdMapBuilder.build();
+            if (!dynamicFilterIdMap.isEmpty()) {
+                // update dynamic filter ids on the probe side
```

Comment:
even if it isn't exponential. It will still be `O(n^2)` where `n` is number of nodes.
Please do a full single `DynamicFilterVisitor` run within `io.trino.sql.planner.optimizations.UnaliasSymbolReferences#optimize` once mappings are collected. This way complexity will be `O(n)`

Reply:
It's not that straightforward. Constants matter. 
I think the better approximation would be `0(n*k)` where `k` is the number of joins that have duplicated dynamic filters and doing the pass in the `optimize` method adds overhead as even the empty df id map has to be propagated to the root. This propagation involves merging multiple source maps in some node types so it's not totally free.
So for small `k` (0, 1, 2) I think doing the pass here is more efficient but of course, asymptotically doing the pass in the `optimize` is better.

Reply:
I added a commit with the single pass. PTAL. 
It passes the `dynamicFilterIdMap` in the `PlanAndMappings` which is the pattern used in trino.
I could avoid most of the changes and the potential perf impact by using mutable state in the `UnaliasSymbolReferences.Visitor` (keep mutable `dynamicFilterIdMap` as a field there)